mdagent/tools/base_tools/simulation_tools/setup_and_run.py

```diff
@@ -873,40 +893,144 @@ def _create_system(
+                        except Exception as e:
+                            print("Error adding solvent", type(e).__name__, "–", e)
+                            raise (e)
+                    else:
+                        print("Exception: ", str(e))
                         raise (e)
-                else:
-                    print("Exception: ", str(e))
-                    raise (e)
 
-            system = forcefield.createSystem(self.modeller.topology, **system_params)
-        else:
-            system = forcefield.createSystem(self.modeller.topology, **system_params)
+                system = forcefield.createSystem(
+                    self.modeller.topology, **system_params
+                )
+                break
+            else:
+                try:
+                    print("adding system without solvent")
+                    system = forcefield.createSystem(
+                        self.modeller.topology, **system_params
+                    )
+                    break
+                except ValueError as e:
+                    if "No template found for" in str(e):
+                        print("Trying to add component to Forcefield...")
+                        pattern = r"residue \d+ \((\w+)\)"
+                        # Search for the pattern in the error message
+                        match = re.search(pattern, str(e))
+                        if match:
+                            residue_code = match.group(1)
+                            print(f"Residue code: {residue_code}")
+                        else:
+                            print("No residue code found in the error message.")
+                            raise ValueError(str(e))
+                        if residue_code not in solvent_list:
+                            print(
+                                "Residue code not in solvent list. Adding forcefield \
+                                  not supported."
+                            )
+                            raise ValueError(str(e))
+                        else:
+                            smiles = self._code_to_smiles(residue_code)
+                            if not smiles:
+                                print("No SMILES found for HET code.")
+                                raise ValueError(str(e))
+
+                            print(f"Found SMILES from HET code: {smiles}")
+
+                            molecule = Molecule.from_smiles(smiles)
+                            smirnoff = SMIRNOFFTemplateGenerator(molecules=molecule)
+                            forcefield.registerTemplateGenerator(smirnoff.generator)
+                            attempts += 1
+                            print(
+                                f"Attempt {attempts} to add small \
+                                  molecules to forcefield."
+                            )
+                            continue
+                    else:
+                        raise ValueError(str(e))
```

Comment:
Looks like these lines are identical to line 905-942. If the approach to ValueError is same for both `if solvate` and `else` conditions, then what's the point of typing them out twice (~40 lines)? Can you write it once somewhere? Also, are force fields identical for standalone small molecules vs. solvent molecules? It seems like the only difference is this one line, `self.modeller.addSolvent(forcefield)`.

Reply:
the difference is that in the .addSolvent is used if the simulation is with water. The other one is if is done in vacuum. 
yeah, I could condense the workflow, but at the beginning I wasn't sure if I was going to end up using exactly the same code, or if one of the options could require a different approach